ionoscloud/resource_dns_zone.go

```diff
@@ -0,0 +1,147 @@
+)
+
+func resourceDNSZone() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: zoneCreate,
+		ReadContext:   zoneRead,
+		UpdateContext: zoneUpdate,
+		DeleteContext: zoneDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: zoneImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+			"nameservers": {
+				Type:        schema.TypeList,
+				Description: "A list of available name servers.",
+				Computed:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func zoneCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(SdkBundle).DNSClient
+	zoneResponse, _, err := client.CreateZone(ctx, d)
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("an error occured while creating a DNS Zone: %w", err))
+	}
+	if zoneResponse.Metadata.State != nil {
+		if *zoneResponse.Metadata.State == dns.FAILED {
+			// This is a temporary error message since right now the API is not returning errors that we can work with.
+			return diag.FromErr(fmt.Errorf("zone creation has failed, this can happen if the data in the request is not correct, " +
+				"please check again the values defined in the plan"))
+		}
+	}
+	d.SetId(*zoneResponse.Id)
+	return zoneRead(ctx, d, meta)
+}
+
+func zoneRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(SdkBundle).DNSClient
+	zoneId := d.Id()
+	zone, apiResponse, err := client.GetZoneById(ctx, zoneId)
+
+	if err != nil {
+		if apiResponse.HttpNotFound() {
+			log.Printf("[INFO] Could not find zone with ID: %s", zoneId)
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("error while fetching DNS Zone with ID: %s, error: %w", zoneId, err))
+	}
+
+	log.Printf("[INFO] Successfully retrieved DNS Zone with ID: %s: %+v", zoneId, zone)
+
+	if err := client.SetZoneData(d, zone); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func zoneUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(SdkBundle).DNSClient
+	zoneId := d.Id()
+
+	zoneResponse, _, err := client.UpdateZone(ctx, zoneId, d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("an error occured while updating the DNS Zone with ID: %s, error: %w", zoneId, err))
+	}
+	if zoneResponse.Metadata.State != nil {
+		if *zoneResponse.Metadata.State == dns.FAILED {
+			// This is a temporary error message since right now the API is not returning errors that we can work with.
+			return diag.FromErr(fmt.Errorf("zone update has failed, this can happen if the data in the request is not correct, " +
```

Comment:
In this case, the API zone is still going to remain in FAILED state, right? Does throwing an error not kind of hide the fact that something on the cloud has changed after running the plan? I'm not experienced with terraform, just trying to understand what is happening here

Reply:
`I'm not experienced with terraform` no worries

I understand your point and it makes sense, it can actually be an edge case that we I didn't think about, I will check to see the current behavior. I expect something like:

1) We try to create the Zone, we send the request and the zone is created in a `FAILED` state;
2) Because of that, we raise an error, and we are not setting anything in the `Terraform` state, this means that Terraform will not track anything related to this resource;
3) The resource will continue to exist in a `FAILED` state and basically this is not good since Terraform won't track this resource.

But I'll come back with updates.

Reply:
Ok so I did a bit more digging and (please correct any of my points if wrong):
- This is known as "terraform drift" and should be corrected externally (via CLI for example)
- Throwing an error in this case is good, since the terraform plan wasn't successful
- maybe consider updating the error message to let the user know:
     1. that a zone has been created/updated/etc., however it is in state `FAILED`
     2. he needs to take some external action to revert the tf plan's changes

Reply:
@avirtopeanu-ionos this is not really a drift since we are not setting anything in the terraform state. Usually, after you receive a response, you set some values in the state in order to have a mapping between the resource defined in the plan and the actual resurce that exists in the cloud. Also, you can check this:
```
Drift is the term for when the real-world state of your infrastructure differs from the state defined in your configuration. 
```
`Throwing an error in this case is good, since the terraform plan wasn't successful` indeed it is, but my problem is that the resource still exists in cloud, at least this is what I expect.
`maybe consider updating the error message to let the user know: ` this is a possible solution, we will check all possible solutions after I test the behavior. I can't test it RN because there are some problems with the DNS API.

Reply:
@avirtopeanu-ionos I was able to test this again, but it seems like the behavior is different, I tried to create a record in the `FAILED` state, but right now I receive a 400 error response, it seems like they changed the behavior. Let's leave it as it is for now and I will ask them how I can generate a response in which the state is `FAILED` in order to check if a resource is created or not.